ext/updater.go

```diff
@@ -157,21 +157,27 @@ func (u *Updater) StartPolling(b *gotgbot.Bot, opts *PollingOpts) error {
 		}
 	}
 
-	bData, err := u.botMapping.addBot(b, "", "")
+	ctx, closeFn := context.WithCancel(context.Background())
+
+	bData, err := u.botMapping.addPollingBot(b, closeFn)
 	if err != nil {
 		return fmt.Errorf("failed to add bot with long polling: %w", err)
 	}
 
 	go u.Dispatcher.Start(b, bData.updateChan)
-	go u.pollingLoop(bData, reqOpts, v)
+
+	bData.updateWriterControl.Add(1)
+	go func() {
+		// defer, so it gets called even in case of panics.
+		defer bData.updateWriterControl.Done()
+
+		u.pollingLoop(ctx, bData, reqOpts, v)
+	}()
 
 	return nil
 }
 
-func (u *Updater) pollingLoop(bData *botData, opts *gotgbot.RequestOpts, v map[string]string) {
-	bData.updateWriterControl.Add(1)
-	defer bData.updateWriterControl.Done()
-
+func (u *Updater) pollingLoop(ctx context.Context, bData *botData, opts *gotgbot.RequestOpts, v map[string]string) {
```

Comment:
@rusq This should fix it! An edge case for sure, but I can see this being triggered in live systems which are dynamically adding/removing bots.

TL;DR: Race condition caused by the waitgroup being started in a goroutine, rather than before it.

This meant that when the stars (mis)aligned, calling `updater.StartPolling(...)` followed by `updater.Stop()`, could result in `.Wait()` (from the `.Stop()`) before calling `.Add(1)` (from the goroutine in `.StartPolling`); which then causes the race.

The fix makes sure that the `.Add` is done sequentially, thus will always run before the `.Wait`.

Reply:
Concurrency bugs are always the hardest to debug, great break down!